com.unity.render-pipelines.universal/Runtime/RendererFeatures/ScreenSpaceShadows.cs

```diff
@@ -0,0 +1,230 @@
+using System;
+using UnityEngine.Experimental.Rendering;
+
+namespace UnityEngine.Rendering.Universal
+{
+    [Serializable]
+    internal class ScreenSpaceShadowsSettings
+    {
+
+    }
+
+    [DisallowMultipleRendererFeature]
+    internal class ScreenSpaceShadows : ScriptableRendererFeature
+    {
+        // Serialized Fields
+        [SerializeField, HideInInspector] private Shader m_Shader = null;
+        [SerializeField] private ScreenSpaceShadowsSettings m_Settings = new ScreenSpaceShadowsSettings();
+
+        // Private Fields
+        private Material m_Material;
+        private ScreenSpaceShadowsPass m_SSShadowsPass = null;
+        private EnableKeywordPass m_EnableKeywordPass = null;
+        private DisableKeywordPass m_DisableKeywordPass = null;
+
+        // Constants
+        private const string k_ShaderName = "Hidden/Universal Render Pipeline/ScreenSpaceShadows";
+
+        /// <inheritdoc/>
+        public override void Create()
+        {
+            if (m_SSShadowsPass == null)
+                m_SSShadowsPass = new ScreenSpaceShadowsPass();
+            if (m_EnableKeywordPass == null)
+                m_EnableKeywordPass = new EnableKeywordPass();
+            if (m_DisableKeywordPass == null)
+                m_DisableKeywordPass = new DisableKeywordPass();
+
+            LoadMaterial();
+            m_SSShadowsPass.profilerTag = name;
+            m_SSShadowsPass.renderPassEvent = RenderPassEvent.AfterRenderingPrepasses;
+
+            m_EnableKeywordPass.renderPassEvent = RenderPassEvent.BeforeRenderingOpaques;
+            m_DisableKeywordPass.renderPassEvent = RenderPassEvent.AfterRenderingOpaques;
+        }
+
+        /// <inheritdoc/>
+        public override void AddRenderPasses(ScriptableRenderer renderer, ref RenderingData renderingData)
+        {
+            if (m_Material == null)
+            {
+                Debug.LogErrorFormat(
+                    "{0}.AddRenderPasses(): Missing material. {1} render pass will not be added. Check for missing reference in the renderer resources.",
+                    GetType().Name, m_SSShadowsPass.profilerTag);
+                return;
+            }
+
+            bool shouldAdd = m_SSShadowsPass.Setup(m_Settings) && renderingData.shadowData.supportsMainLightShadows;
+            if (shouldAdd)
+            {
+                renderer.EnqueuePass(m_SSShadowsPass);
+                renderer.EnqueuePass(m_EnableKeywordPass);
+                renderer.EnqueuePass(m_DisableKeywordPass);
```

Comment:
Suggestion to make less boiler plate code. You don't necessarily need the enable and disable the keywords by adding passes for it. You can enable the keyword in the SSShadowPass, you can disable the keyword in the SrciptableRenderer before drawing the transparent block.

Reply:
Remove a enabling keyword pass, but it doesn't seem for the position before drawing transparent blocks to be proper, so I added disabling the keyword in transparent settings pass. if so the transparent settings pass must be called in all cases of graphics settings, because the transparent settings pass can't know whether the screen space shadow was called or not.(If we can investigate if any render pass was enqueued or not, it will be useful IMO)

Reply:
If you disable the keyword inside the transparent pass or ScriptableRenderer, doesn't that make this ScriptableRendererFeature not self-contained anymore? Can users actually insert ScriptableRendererFeature in other ScriptableRenderers?

Reply:
Yes, I pulled out those enabling/disabling keyword pass. and in TransparentSettingsPass 'screen space shadow' keyword will be disabled. this TransparentSettingsPass was called in the case of not receiving shadows on transparent objects from 'Setup' function of that, but in my PR the transparent pass will be always called :(